File: code/programs/rust/unix-tools/src/cat_tool.rs
//! # cat — Concatenate and Print Files
//!
//! This module implements the business logic for the `cat` command.
//! The `cat` utility reads files sequentially, writing their contents
//! to standard output. If no file is specified, or if the file is `-`,
//! it reads from standard input.
//!
//! ## Options Overview
//!
//! The GNU version of `cat` supports several display options:
//!
//! ```text
//!     -n       Number all output lines
//!     -b       Number only non-blank lines (overrides -n)
//!     -s       Squeeze consecutive blank lines into one
//!     -T       Show tabs as ^I
//!     -E       Show $ at end of each line
//!     -v       Show non-printing characters using ^ and M- notation
//!     -A       Equivalent to -vET (show all)
//! ```
//!
//! ## Design
//!
//! The core function `process_cat_content` takes a string of content
//! and a set of options, then returns the transformed output. This
//! keeps the I/O (reading files, writing stdout) separate from the
//! transformation logic, making everything testable.

use std::io::{self, Read};
use std::fs;

// ---------------------------------------------------------------------------
// Options struct
// ---------------------------------------------------------------------------

/// Configuration options for cat output transformation.
///
/// These correspond directly to the command-line flags defined in
/// `cat.json`. The `show_all` flag is expanded into its component
/// parts (`show_ends`, `show_tabs`, `show_nonprinting`) before being
/// stored here.
#[derive(Debug, Clone)]
pub struct CatOptions {
    /// Number all output lines (`-n`)
    pub number: bool,
    /// Number only non-blank lines, overrides `number` (`-b`)
    pub number_nonblank: bool,
    /// Squeeze consecutive blank lines into one (`-s`)
    pub squeeze_blank: bool,
    /// Display TAB characters as ^I (`-T`)
    pub show_tabs: bool,
    /// Display $ at end of each line (`-E`)
    pub show_ends: bool,
    /// Show non-printing characters in ^ and M- notation (`-v`)
    pub show_nonprinting: bool,
}

impl CatOptions {
    /// Create a new `CatOptions` with all options disabled.
    pub fn new() -> Self {
        CatOptions {
            number: false,
            number_nonblank: false,
            squeeze_blank: false,
            show_tabs: false,
            show_ends: false,
            show_nonprinting: false,
        }
    }

    /// Apply the `show_all` (`-A`) flag by enabling `-v`, `-E`, and `-T`.
    ///
    /// The `-A` flag is syntactic sugar — it's equivalent to passing
    /// all three flags individually. We expand it here so the rest of
    /// the code doesn't need to know about `-A`.
    pub fn apply_show_all(&mut self) {
        self.show_nonprinting = true;
        self.show_ends = true;
        self.show_tabs = true;
    }
}

// ---------------------------------------------------------------------------
// Core transformation
// ---------------------------------------------------------------------------

/// Transform file content according to the given cat options.
///
/// This is the pure business logic function. It takes a string of content
/// and returns the transformed output. No I/O happens here.
///
/// ## Line Processing Pipeline
///
/// Each line flows through these stages:
///
/// ```text
///     Input line
///       │
///       ├─► Squeeze blank?  (skip if consecutive blank and -s)
///       ├─► Show non-printing chars?  (transform if -v)
///       ├─► Show tabs?  (replace \t with ^I if -T)
///       ├─► Show ends?  (append $ if -E)
///       └─► Number?  (prepend line number if -n or -b)
///           │
///           ▼
///     Output line
/// ```
pub fn process_cat_content(content: &str, options: &CatOptions) -> String {
    let mut output = String::with_capacity(content.len());
    let mut line_number: usize = 1;
    let mut prev_blank = false;

    // We split on newline boundaries but preserve the structure.
    // Using `.split('\n')` gives us the lines without their terminators.
    // We need to re-add them (except possibly for the very last one
    // if the input doesn't end with a newline).
    let lines: Vec<&str> = content.split('\n').collect();
    let last_idx = if lines.last() == Some(&"") {
        // Content ended with \n — the split produces an empty trailing
        // element that we should not treat as a real line.
        lines.len() - 1
    } else {
        lines.len()
    };

    for (i, line) in lines.iter().enumerate() {
        if i >= last_idx {
            // This is the phantom empty string after a trailing newline.
            break;
        }

        let is_blank = line.is_empty();

        // --- Squeeze blank lines ---
        // If -s is set and we see consecutive blank lines, skip all but
        // the first one. This collapses runs of empty lines into a single
        // empty line.
        if options.squeeze_blank && is_blank && prev_blank {
            continue;
        }
        prev_blank = is_blank;

        // --- Build the transformed line ---
        let mut transformed = String::new();

        // --- Line numbering ---
        // -b (number_nonblank) takes precedence over -n (number).
        // With -b, blank lines are not numbered.
        // With -n, all lines are numbered.
        if options.number_nonblank {
            if !is_blank {
                transformed.push_str(&format!("{:>6}\t", line_number));
                line_number += 1;
            } else {
                // Blank lines get no number, just the empty line
            }
        } else if options.number {
            transformed.push_str(&format!("{:>6}\t", line_number));
            line_number += 1;
        }

        // --- Transform the line content ---
        for ch in line.chars() {
            if options.show_tabs && ch == '\t' {
                // Show tabs as ^I (caret notation)
                transformed.push_str("^I");
            } else if options.show_nonprinting && ch != '\t' && ch != '\n' {
                // Show non-printing characters using ^ and M- notation.
                // Tabs and newlines are excluded from this transformation
                // (they have their own flags).
                transformed.push_str(&format_nonprinting(ch));
            } else {
                transformed.push(ch);
            }
        }

        // --- Show line endings ---
        if options.show_ends {
            transformed.push('$');
        }

        // --- Append newline ---
        transformed.push('\n');
        output.push_str(&transformed);
    }

    output
}

/// Read content from a file path, or from stdin if path is "-".
///
/// This handles the I/O side of cat. The result is a string containing
/// the file's entire contents, ready for transformation by
/// `process_cat_content`.
pub fn read_file_or_stdin(path: &str) -> Result<String, String> {
    if path == "-" {
        // Read from standard input
        let mut buffer = String::new();
        io::stdin()
            .lock()
            .read_to_string(&mut buffer)
            .map_err(|e| format!("cat: -: {}", e))?;
        Ok(buffer)
    } else {
        // Read from a file
        fs::read_to_string(path)
            .map_err(|e| format!("cat: {}: {}", path, e))
    }
}

// ---------------------------------------------------------------------------
// Non-printing character formatting
// ---------------------------------------------------------------------------

/// Format a non-printing character using ^ and M- notation.
///
/// This implements the same notation used by GNU `cat -v`:
///
/// ```text
///     Byte range     Notation         Example
///     ─────────────  ───────────────  ──────────────
///     0x00 – 0x1F    ^@, ^A, ..., ^_ ^@ for NUL
///     0x7F           ^?               DEL character
///     0x80 – 0x9F    M-^@, M-^A, ... High-bit control
///     0xA0 – 0xFE    M- , M-!, ...   High-bit printable
///     0xFF           M-^?             High-bit DEL
/// ```
///
/// For multi-byte UTF-8 characters (code points > 255), we simply
/// pass them through unchanged — they're printable Unicode.
fn format_nonprinting(ch: char) -> String {
    let code = ch as u32;

    if code > 255 {
        // Multi-byte Unicode character — pass through as-is.
        // GNU cat -v does the same for valid UTF-8 sequences.
        return ch.to_string();
    }

    let byte = code as u8;

    if byte < 32 {
        // Control characters (0x00-0x1F): use ^X notation
        // ^@ = 0, ^A = 1, ..., ^Z = 26, ^[ = 27, ...
        format!("^{}", (byte + 64) as char)
    } else if byte == 127 {
        // DEL character: ^?
        "^?".to_string()
    } else if byte >= 128 && byte < 160 {
        // High-bit control characters: M-^X notation
        format!("M-^{}", (byte - 128 + 64) as char)
    } else if byte >= 160 && byte < 255 {
        // High-bit printable characters: M-X notation
        format!("M-{}", (byte - 128) as char)
    } else if byte == 255 {
        // 0xFF: M-^?
        "M-^?".to_string()
    } else {
        // Regular printable ASCII (32-126): pass through
        ch.to_string()
    }
}

// ---------------------------------------------------------------------------
// Unit tests
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn simple_passthrough() {
        let opts = CatOptions::new();
        assert_eq!(
            process_cat_content("hello\nworld\n", &opts),
            "hello\nworld\n"
        );
    }

    #[test]
    fn number_all_lines() {
        let mut opts = CatOptions::new();
        opts.number = true;
        let result = process_cat_content("a\nb\nc\n", &opts);
        assert!(result.contains("     1\ta"));
        assert!(result.contains("     2\tb"));
        assert!(result.contains("     3\tc"));
    }

    #[test]
    fn number_nonblank_skips_empty() {
        let mut opts = CatOptions::new();
        opts.number_nonblank = true;
        let result = process_cat_content("a\n\nb\n", &opts);
        // Line "a" gets number 1, blank line gets no number, "b" gets 2
        assert!(result.contains("     1\ta"));
        assert!(result.contains("     2\tb"));
        // The blank line should not have a number
        let lines: Vec<&str> = result.lines().collect();
        assert_eq!(lines[1], ""); // blank line, no number prefix
    }

    #[test]
    fn squeeze_blank_lines() {
        let mut opts = CatOptions::new();
        opts.squeeze_blank = true;
        let result = process_cat_content("a\n\n\n\nb\n", &opts);
        // Multiple blank lines should be squeezed to one
        assert_eq!(result, "a\n\nb\n");
    }

    #[test]
    fn show_tabs() {
        let mut opts = CatOptions::new();
        opts.show_tabs = true;
        let result = process_cat_content("hello\tworld\n", &opts);
        assert_eq!(result, "hello^Iworld\n");
    }

    #[test]
    fn show_ends() {
        let mut opts = CatOptions::new();
        opts.show_ends = true;
        let result = process_cat_content("hello\nworld\n", &opts);
        assert_eq!(result, "hello$\nworld$\n");
    }

    #[test]
    fn show_all_enables_three_flags() {
        let mut opts = CatOptions::new();
        opts.apply_show_all();
        assert!(opts.show_nonprinting);
        assert!(opts.show_ends);
        assert!(opts.show_tabs);
    }

    #[test]
    fn nonprinting_control_chars() {
        assert_eq!(format_nonprinting('\x01'), "^A");
        assert_eq!(format_nonprinting('\x00'), "^@");
        assert_eq!(format_nonprinting('\x1F'), "^_");
        assert_eq!(format_nonprinting('\x7F'), "^?");
    }

    #[test]
    fn empty_input() {
        let opts = CatOptions::new();
        assert_eq!(process_cat_content("", &opts), "");
    }

    #[test]
    fn no_trailing_newline() {
        let opts = CatOptions::new();
        // Input without trailing newline — we still process the last line
        assert_eq!(process_cat_content("hello\n", &opts), "hello\n");
    }
}

File: code/programs/go/unix-tools/cat_tool.go
// =========================================================================
// cat — Concatenate Files and Print to Standard Output
// =========================================================================
//
// The `cat` utility reads files sequentially and writes their contents
// to standard output. Its name comes from "concatenate" — its primary
// purpose is joining multiple files together.
//
// # Basic usage
//
//   cat file.txt                Read one file
//   cat file1.txt file2.txt     Concatenate two files
//   cat                         Read from standard input (stdin)
//   cat -                       Same as above (explicit stdin)
//   cat file1.txt - file2.txt   Mix files and stdin
//
// # Flags
//
// cat has several display-modifying flags. None of them change the actual
// content — they only affect how it appears on screen:
//
//   -n    Number all output lines (starting from 1).
//   -b    Number only non-blank lines (overrides -n).
//   -s    Squeeze repeated blank lines into a single blank line.
//   -T    Show tab characters as ^I (caret notation).
//   -E    Show a $ at the end of each line (makes trailing spaces visible).
//   -v    Show non-printing characters using ^ and M- notation.
//   -A    Show all: equivalent to -vET.
//
// # How line numbering works
//
//	Input:           -n output:        -b output:
//	┌────────────┐   ┌──────────────┐  ┌──────────────┐
//	│ hello      │   │ 1  hello     │  │ 1  hello     │
//	│            │   │ 2            │  │              │
//	│ world      │   │ 3  world     │  │ 2  world     │
//	└────────────┘   └──────────────┘  └──────────────┘
//
// Notice that -b skips numbering the blank line (line 2 in -n mode).
// The -b flag overrides -n: if both are given, -b wins.
//
// # Squeeze blank lines (-s)
//
//	Input:           -s output:
//	┌────────────┐   ┌────────────┐
//	│ hello      │   │ hello      │
//	│            │   │            │
//	│            │   │ world      │
//	│            │   │            │
//	│ world      │   └────────────┘
//	└────────────┘
//
// Multiple consecutive blank lines are collapsed into one.
//
// # Architecture
//
//	cat.json (spec)          cat_tool.go (this file)
//	┌──────────────────┐     ┌──────────────────────────────┐
//	│ flags: -n,-b,-s  │     │ for each file:               │
//	│ -T,-E,-v,-A      │────>│   read lines                 │
//	│ variadic FILE    │     │   apply transforms            │
//	│ help, version    │     │   write to stdout             │
//	└──────────────────┘     └──────────────────────────────┘

package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"strings"

	clibuilder "github.com/adhithyan15/coding-adventures/code/packages/go/cli-builder"
)

// =========================================================================
// catOptions — configuration for the cat operation
// =========================================================================
//
// We bundle all the boolean flags into a struct for clarity. This avoids
// passing 7 separate booleans to functions, which would be error-prone
// and hard to read.

type catOptions struct {
	number        bool // -n: number all lines
	numberNonblank bool // -b: number only non-blank lines (overrides -n)
	squeezeBlank  bool // -s: squeeze consecutive blank lines
	showTabs      bool // -T: display tabs as ^I
	showEnds      bool // -E: display $ at end of each line
	showNonprint  bool // -v: show non-printing characters
}

// =========================================================================
// showNonPrintingChar — convert a byte to its ^ or M- representation
// =========================================================================
//
// Non-printing characters (control characters, DEL, and characters with
// the high bit set) are displayed using a notation that makes them visible:
//
//   Control characters (0x00-0x1F): ^@ through ^_ (except TAB and LFD)
//   DEL (0x7F): ^?
//   High-bit characters (0x80-0xFF): M-<char> where <char> is the
//     low-7-bit representation
//
// This is the same notation used by `cat -v` in GNU coreutils.
//
// Examples:
//   0x00 (NUL)  => ^@
//   0x01 (SOH)  => ^A
//   0x1B (ESC)  => ^[
//   0x7F (DEL)  => ^?
//   0x80        => M-^@
//   0xC0        => M-@

func showNonPrintingChar(b byte) string {
	// Characters with the high bit set (128-255) get the M- prefix.
	// We strip the high bit, process the low 7 bits, then prepend M-.
	if b >= 128 {
		// Strip the high bit and recurse on the low 7 bits.
		if b == 128+127 {
			return "M-^?"
		}
		low := b - 128
		if low < 32 {
			return fmt.Sprintf("M-^%c", low+'@')
		}
		return fmt.Sprintf("M-%c", low)
	}

	// DEL character (127) is shown as ^?
	if b == 127 {
		return "^?"
	}

	// Control characters (0-31) are shown as ^@ through ^_
	// Exception: TAB (9) and LFD/newline (10) are NOT converted —
	// they're handled separately by -T and normal line processing.
	if b < 32 && b != '\t' && b != '\n' {
		return fmt.Sprintf("^%c", b+'@')
	}

	// Regular printable character — return as-is.
	return string(b)
}

// =========================================================================
// processLine — apply display transforms to a single line
// =========================================================================
//
// This function takes a raw line (without the trailing newline) and
// applies the requested display transforms in the correct order:
//   1. Show non-printing characters (-v)
//   2. Show tabs as ^I (-T)
//   3. Show line ending as $ (-E)
//
// The transforms are applied per-character, building a new string.

func processLine(line string, opts catOptions) string {
	// Fast path: if no character-level transforms are needed, return as-is.
	if !opts.showNonprint && !opts.showTabs && !opts.showEnds {
		return line
	}

	var result strings.Builder

	for i := 0; i < len(line); i++ {
		b := line[i]

		// Handle TAB characters.
		if b == '\t' {
			if opts.showTabs {
				// Show tabs as ^I (caret notation for horizontal tab).
				// The "I" comes from: TAB is the 9th character (0-indexed),
				// and '@' + 9 = 'I' in ASCII.
				result.WriteString("^I")
			} else {
				result.WriteByte('\t')
			}
			continue
		}

		// Handle non-printing characters.
		if opts.showNonprint && (b < 32 || b >= 127) && b != '\t' && b != '\n' {
			result.WriteString(showNonPrintingChar(b))
			continue
		}

		// Regular character — pass through.
		result.WriteByte(b)
	}

	// Append $ at the end of the line if -E is set.
	if opts.showEnds {
		result.WriteByte('$')
	}

	return result.String()
}

// =========================================================================
// catReader — process a single reader (file or stdin) with cat options
// =========================================================================
//
// This function reads from the given reader line by line and writes to
// stdout with the requested transforms applied. The lineNum pointer
// allows line numbering to continue across multiple files (cat file1 file2
// should have continuous line numbers).
//
// Parameters:
//   - reader: the input source (file or stdin)
//   - stdout: where to write the output
//   - opts: which display transforms to apply
//   - lineNum: pointer to the current line number (shared across files)
//   - prevBlank: pointer to whether the previous line was blank (for -s)

func catReader(reader io.Reader, stdout io.Writer, opts catOptions, lineNum *int, prevBlank *bool) error {
	scanner := bufio.NewScanner(reader)

	// Increase the scanner buffer size to handle very long lines.
	// The default is 64KB, which is usually enough, but some files
	// (like minified JavaScript) can have lines much longer than that.
	buf := make([]byte, 0, 1024*1024)
	scanner.Buffer(buf, 1024*1024)

	for scanner.Scan() {
		line := scanner.Text()
		isBlank := len(line) == 0

		// Handle -s (squeeze blank lines).
		// If this line is blank and the previous line was also blank,
		// skip this line entirely.
		if opts.squeezeBlank && isBlank && *prevBlank {
			continue
		}
		*prevBlank = isBlank

		// Apply character-level transforms to the line content.
		processed := processLine(line, opts)

		// Handle line numbering.
		if opts.numberNonblank {
			// -b: number only non-blank lines.
			if !isBlank {
				*lineNum++
				fmt.Fprintf(stdout, "%6d\t%s\n", *lineNum, processed)
			} else {
				fmt.Fprintln(stdout, processed)
			}
		} else if opts.number {
			// -n: number all lines.
			*lineNum++
			fmt.Fprintf(stdout, "%6d\t%s\n", *lineNum, processed)
		} else {
			// No numbering — just print the line.
			fmt.Fprintln(stdout, processed)
		}
	}

	return scanner.Err()
}

// =========================================================================
// runCat — the testable core of the cat tool
// =========================================================================
//
// The cat tool processes each file in order, applying the requested
// display transforms. If no files are specified, it reads from stdin.
// The special filename "-" also means stdin.
//
// Line numbering continues across files: if file1 has 3 lines and file2
// has 2 lines, the total numbering goes 1-5.

func runCat(specPath string, argv []string, stdout io.Writer, stderr io.Writer) int {
	return runCatWithStdin(specPath, argv, stdout, stderr, os.Stdin)
}

// runCatWithStdin is the inner implementation that accepts a custom stdin
// reader. This allows tests to provide mock stdin input without modifying
// the real os.Stdin.

func runCatWithStdin(specPath string, argv []string, stdout io.Writer, stderr io.Writer, stdin io.Reader) int {
	// Step 1: Create the parser from the spec.
	parser, err := clibuilder.NewParser(specPath, argv)
	if err != nil {
		fmt.Fprintf(stderr, "cat: %s\n", err)
		return 1
	}

	// Step 2: Parse the arguments.
	result, err := parser.Parse()
	if err != nil {
		fmt.Fprintf(stderr, "%s\n", err)
		return 1
	}

	// Step 3: Handle the result.
	switch r := result.(type) {

	case *clibuilder.HelpResult:
		fmt.Fprintln(stdout, r.Text)
		return 0

	case *clibuilder.VersionResult:
		fmt.Fprintln(stdout, r.Version)
		return 0

	case *clibuilder.ParseResult:
		// Extract flags into our options struct.
		opts := catOptions{
			number:        getBool(r.Flags, "number"),
			numberNonblank: getBool(r.Flags, "number_nonblank"),
			squeezeBlank:  getBool(r.Flags, "squeeze_blank"),
			showTabs:      getBool(r.Flags, "show_tabs"),
			showEnds:      getBool(r.Flags, "show_ends"),
			showNonprint:  getBool(r.Flags, "show_nonprinting"),
		}

		// -A (show all) is equivalent to -vET.
		if getBool(r.Flags, "show_all") {
			opts.showNonprint = true
			opts.showEnds = true
			opts.showTabs = true
		}

		// -b overrides -n. If both are set, only number non-blank lines.
		if opts.numberNonblank {
			opts.number = false
		}

		// Extract file paths from arguments.
		files := getStringSlice(r.Arguments, "files")

		// If no files specified, default to stdin.
		if len(files) == 0 {
			files = []string{"-"}
		}

		// Process each file in order.
		lineNum := 0
		prevBlank := false
		exitCode := 0

		for _, file := range files {
			var reader io.Reader

			if file == "-" {
				// "-" means read from standard input.
				reader = stdin
			} else {
				// Open the file for reading.
				f, err := os.Open(file)
				if err != nil {
					fmt.Fprintf(stderr, "cat: %s: %s\n", file, err)
					exitCode = 1
					continue
				}
				defer f.Close()
				reader = f
			}

			if err := catReader(reader, stdout, opts, &lineNum, &prevBlank); err != nil {
				fmt.Fprintf(stderr, "cat: %s: %s\n", file, err)
				exitCode = 1
			}
		}

		return exitCode

	default:
		fmt.Fprintf(stderr, "cat: unexpected result type: %T\n", result)
		return 1
	}
}

// =========================================================================
// Helper functions for extracting typed values from parsed results
// =========================================================================

// getBool safely extracts a boolean value from a map.
// Returns false if the key doesn't exist or the value isn't a bool.
func getBool(m map[string]interface{}, key string) bool {
	v, ok := m[key].(bool)
	if !ok {
		return false
	}
	return v
}

// getStringSlice extracts a string slice from a parsed arguments map.
//
// cli-builder returns variadic arguments as []interface{}, where each
// element is a string. This function handles the type conversion safely.
func getStringSlice(m map[string]interface{}, key string) []string {
	raw, ok := m[key]
	if !ok {
		return nil
	}

	// Try []interface{} first (what cli-builder returns for variadic args).
	if slice, ok := raw.([]interface{}); ok {
		result := make([]string, 0, len(slice))
		for _, v := range slice {
			if s, ok := v.(string); ok {
				result = append(result, s)
			}
		}
		return result
	}

	// Try a single string (for non-variadic args or single values).
	if s, ok := raw.(string); ok {
		return []string{s}
	}

	return nil
}
